src/dualnumber.jl

```diff
@@ -161,6 +159,8 @@ Base.convert(::Type{Dual{T,V,P}}, d::Dual{T}) where {T,V,P} = Dual{T}(convert(V,
 Base.convert(::Type{Dual{T,V,P}}, x) where {T,V,P} = Dual{T}(convert(V, x), zero(P))
 Base.convert(::Type{Dual{T,V,P}}, x::Number) where {T,V,P} = Dual{T}(convert(V, x), zero(P))
 Base.convert(::Type{D}, d::D) where {D<:Dual} = d
+Base.float(d::Dual{T,V,P}) where {T,V,P} = convert(Dual{T,float(V),P}, d)
+Base.AbstractFloat(d::Dual) = float(d)
```

Comment:
What's this line for?
Does it match to base behavior? e.g. on Complex?

Reply:
https://github.com/YingboMa/ForwardDiff2.jl/pull/19/commits/58ededfcdf7af5e01a6446960d16704351286020

Reply:
We don't care about Complex right now.

Reply:
The problem is about API.
Generally when `Foo` is a type, `Foo(x)::Foo` holds always.
This breaks it.

Reply:
```julia
julia> AbstractFloat(ForwardDiff.Dual(1, 2))::AbstractFloat
ERROR: TypeError: in typeassert, expected AbstractFloat, got ForwardDiff.Dual{Nothing,Float16,1}
Stacktrace:
 [1] top-level scope at REPL[37]:1
```
ForwardDiff doesn't handle it either, and there is nothing we can do about it.

Reply:
My question is: why do we need to define `Base.AbstractFloat(Dual)` in the first place?

Reply:
It makes AD work more code, so why not.

Reply:
'k.
I have never seen `AbstractFloat` used this way in the wild.
Only `float`.